crates/adapters/src/format/avro/deserializer.rs

```diff
@@ -41,211 +46,112 @@ use std::{
 
 use super::input::avro_de_config;
 
+fn deserialize_decimal<E: serde::de::Error>(d: &Decimal, schema: &Schema) -> Result<String, E> {
+    let Schema::Decimal(schema) = schema else {
+        return Err(E::custom(format!(
+            "expected Decimal schema, but found {schema:?}"
+        )));
+    };
+
+    // TODO: this is really expensive, and can be optimized by passing a binary representation of
+    // mantissa + scale. This will require a customer `DeserializeWithContext` implementation for
+    // rust_decimal::Decimal.
+    let bigdecimal = BigDecimal::new(BigInt::from(d.clone()), schema.scale as i64);
+    Ok(bigdecimal.to_string())
+}
+
 pub struct Deserializer<'de> {
+    // We have to keep the schema around, since `Value` isn't sufficiently self-descriptive
+    // to deserialize from it. Specifically, `Decimal` only stores the mantissa; scale must be
+    // extracted from the schema.
+    schema: &'de Schema,
     input: &'de Value,
 }
 
 struct SeqDeserializer<'de> {
     input: Iter<'de, Value>,
+    item_schema: &'de Schema,
 }
 
 struct MapDeserializer<'de> {
+    value_schema: &'de Schema,
```

Comment:
only value?

Reply:
yeah, in Avro maps must have strings as keys.